### PR TITLE
Sidebar reacts to screen size and refactor to use interface package

### PR DESCRIPTION
**Pull request by jorgefilipecosta**
This PR applies the following changes:
- Implements the functionality to automatically close the sidebar when the screen is resized from desktop and reopened when the opposite happens.
- Refactors the edit-post module to use the mobile sidebar control from the interface package.
- Refactors edit-post to use the sidebar toggle button the interface package instead of repeating its own.
- Makes edit site and edit widgets use the mobile sidebar control behavior from interface package. Currently it was impossible to open the block inspector on mobile in both screens.
- Implements the sidebars e.g: block inspector on the customizer blocks.

### CODE REVIEW

File: packages/edit-post/src/components/header/index.js [hunk 1]
```diff
@@ -25,47 +22,20 @@ function Header( {
...
 			isPublishSidebarOpened: select(
 				'core/edit-post'
 			).isPublishSidebarOpened(),
 			isSaving: select( 'core/edit-post' ).isSavingMetaBoxes(),
-			getBlockSelectionStart: select( 'core/block-editor' )
-				.getBlockSelectionStart,
 			isPostSaveable: select( 'core/editor' ).isEditedPostSaveable(),
 			deviceType: select(
 				'core/edit-post'
 			).__experimentalGetPreviewDeviceType(),
```
Comment (epiqueras) [inline packages/edit-post/src/components/header/index.js:35]: I don't see these being used.

File: packages/edit-post/src/components/layout/index.js [hunk 2]
```diff
@@ -154,6 +146,7 @@ function Layout() {
 			<LocalAutosaveMonitor />
 			<EditPostKeyboardShortcuts />
 			<EditorKeyboardShortcutsRegister />
+			<SettingsSidebar />
```
Comment (epiqueras) [inline packages/edit-post/src/components/layout/index.js:149]: Why was this moved here?

File: packages/edit-post/src/components/sidebar/settings-sidebar/index.js [hunk 3]
```diff
@@ -22,25 +23,33 @@ import { __ } from '@wordpress/i18n';
...
+			'core/edit-post'
+		);
+		if (
+			! [ 'edit-post/document', 'edit-post/block' ].includes( sidebar )
+		) {
+			if ( select( 'core/block-editor' ).getBlockSelectionStart() ) {
+				sidebar = 'edit-post/block';
+			}
+			sidebar = 'edit-post/document';
+		}
```
Comment (epiqueras) [inline packages/edit-post/src/components/sidebar/settings-sidebar/index.js:37]: This deserves a comment. Why do we now have a fallback instead of returning null?

File: packages/edit-site/src/components/editor/index.js [hunk 4]
```diff
@@ -135,6 +146,7 @@ function Editor( { settings: _settings } ) {
 								<Context.Provider value={ context }>
 									<FocusReturnProvider>
 										<KeyboardShortcuts.Register />
+										<Sidebar />
```
Comment (epiqueras) [inline packages/edit-site/src/components/editor/index.js:149]: Why is this sidebar out here now and how is it different from the new one below?

File: packages/edit-widgets/src/components/customizer-edit-widgets-initializer/style.scss [hunk 5]
```diff
@@ -10,4 +11,15 @@
...
+	.edit-widgets-sidebar {
+		position: absolute;
+		top: 0;
+		margin-left: -12px;
+		margin-right: -12px;
+		width: calc(100% + 24px);
+	}
+
+	.edit-widgets-sidebar {
+		z-index: z-index(".edit-widgets-sidebar");
```
Comment (epiqueras) [inline packages/edit-widgets/src/components/customizer-edit-widgets-initializer/style.scss:23]: These can be merged.

File: packages/edit-site/src/components/editor/index.js (hunk 4 above)
Comment (epiqueras) [reply packages/edit-site/src/components/editor/index.js:149]: Does this have the fills for the slot below?

Comment (epiqueras) [reply packages/edit-site/src/components/editor/index.js:149]: Why is that abstraction necessary?

File: packages/interface/src/components/complementary-area/index.js [hunk 6]
```diff
@@ -32,6 +33,53 @@ function ComplementaryAreaFill( { scope, children, className } ) {
...
+	const shouldOpenWhenNotSmall = useRef( false );
+	const { enableComplementaryArea, disableComplementaryArea } = useDispatch(
+		'core/interface'
+	);
+	useEffect( () => {
+		// If the complementary area is active and the editor is switching from a big to a small window size.
+		if ( isActive && isSmall && ! previousIsSmall.current ) {
+			// Disable the complementary area.
+			disableComplementaryArea( scope );
+			// Flag the complementary area to be reoppen when the window size goes from small to big.
```
Comment (epiqueras) [inline packages/interface/src/components/complementary-area/index.js:53]:
```suggestion
			// Flag the complementary area to be reopened when the window size goes from small to big.
```

File: packages/interface/src/components/complementary-area/index.js [hunk 7]
```diff
@@ -32,6 +33,53 @@ function ComplementaryAreaFill( { scope, children, className } ) {
...
+			shouldOpenWhenNotSmall.current = true;
+		} else if (
+			// If there is a flag indicating the complementary area should be enable when we go from small to big window size
```
Comment (epiqueras) [inline packages/interface/src/components/complementary-area/index.js:56]:
```suggestion
			// If there is a flag indicating the complementary area should be enabled when we go from small to big window size
```

File: packages/interface/src/components/complementary-area/index.js [hunk 8]
```diff
@@ -32,6 +33,53 @@ function ComplementaryAreaFill( { scope, children, className } ) {
...
+			// and we are going from a small to big window size.
+			shouldOpenWhenNotSmall.current &&
+			! isSmall &&
+			previousIsSmall.current
+		) {
+			// Remove the flag indicating the complementary area should be enable.
```
Comment (epiqueras) [inline packages/interface/src/components/complementary-area/index.js:62]:
```suggestion
			// Remove the flag indicating the complementary area should be enabled.
```

File: packages/interface/src/components/complementary-area/index.js [hunk 9]
```diff
@@ -32,6 +33,53 @@ function ComplementaryAreaFill( { scope, children, className } ) {
...
+			shouldOpenWhenNotSmall.current = false;
+			// Enable the complementary area.
+			enableComplementaryArea( scope, identifier );
+		} else if (
+			// If the is flag indicating the current complementary should be reopen but another complementary area becomes active
```
Comment (epiqueras) [inline packages/interface/src/components/complementary-area/index.js:67]:
```suggestion
			// If the flag is indicating the current complementary should be reopened but another complementary area becomes active,
```

File: packages/interface/src/components/complementary-area/index.js [hunk 10]
```diff
@@ -32,6 +33,53 @@ function ComplementaryAreaFill( { scope, children, className } ) {
...
+			activeArea !== identifier
+		) {
+			shouldOpenWhenNotSmall.current = false;
+		}
+		if ( isSmall !== previousIsSmall.current ) {
+			previousIsSmall.current = isSmall;
+		}
+	}, [ isActive, isSmall, scope, identifier, activeArea ] );
+
+	useEffect( () => {}, [ identifier, activeArea ] );
```
Comment (epiqueras) [inline packages/interface/src/components/complementary-area/index.js:80]: What is this for?

File: packages/edit-site/src/components/editor/index.js (hunk 4 above)
Comment (epiqueras) [reply packages/edit-site/src/components/editor/index.js:149]: A name like `SidebarComplementaryAreaFills` would make it easier to grasp.

File: packages/edit-post/src/components/layout/index.js (hunk 2 above)
Comment (jorgefilipecosta) [reply packages/edit-post/src/components/layout/index.js:149]: Our slot mechanism for complementary areas works with assumption multiple fills are rendered. Only the fill of the active complementary area is shown, but buttons to toggle the complementary area are rendered for all the areas.
Previously the core sidebar fill was only rendered when the sidebar was enabled, and we had a custom made toggle button for the sidebar.
This PR makes sure the core sidebars use the same logic as the plugin sidebars so we need to move SettingsSidebar to a position where it is unconditionoly rendered even if no sidebars are active. Otherwise, when no sidebars are active, we would not render the toggle button for core sidebars.

**Review by jorgefilipecosta — COMMENTED** [review]
Hi @epiqueras thank you for the review. Your feedback was applied!

File: packages/edit-post/src/components/sidebar/settings-sidebar/index.js [hunk 11]
```diff
@@ -22,25 +23,40 @@ import { __ } from '@wordpress/i18n';
...
-			),
-		[]
-	);
-	if (
-		! [ 'edit-post/document', 'edit-post/block' ].includes( sidebarName )
-	) {
-		return null;
-	}
+	const { sidebarName, keyboardShortcut } = useSelect( ( select ) => {
+		// The settings sidebar is used by edit-post/document and edit-post/block sidebars.
```
Comment (epiqueras) [inline packages/edit-post/src/components/sidebar/settings-sidebar/index.js:27]:
```suggestion
		// The settings sidebar is used by the edit-post/document and edit-post/block sidebars.
```

File: packages/edit-post/src/components/sidebar/settings-sidebar/index.js [hunk 12]
```diff
@@ -22,25 +23,40 @@ import { __ } from '@wordpress/i18n';
...
+		// sidebarName represents the sidebar that is active or that should be active when SettingsSidebar toggle button is pressed.
```
Comment (epiqueras) [inline packages/edit-post/src/components/sidebar/settings-sidebar/index.js:28]:
```suggestion
		// sidebarName represents the sidebar that is active or that should be active when the SettingsSidebar toggle button is pressed.
```

File: packages/edit-post/src/components/sidebar/settings-sidebar/index.js [hunk 13]
```diff
@@ -22,25 +23,40 @@ import { __ } from '@wordpress/i18n';
...
+		// If one of the two sidebars is active the component will contain the content of that sidebar.
+		// When none of the the two sidebars is not active we can not simply return null, because PluginSidebarEditPost
```
Comment (epiqueras) [inline packages/edit-post/src/components/sidebar/settings-sidebar/index.js:30]:
```suggestion
		// When neither of the the two sidebars is active we can not simply return null, because the PluginSidebarEditPost
```

File: packages/edit-post/src/components/sidebar/settings-sidebar/index.js [hunk 14]
```diff
@@ -22,25 +23,40 @@ import { __ } from '@wordpress/i18n';
...
+		// component besides being used to render the sidebar also renders the toggle button. In that case sidebarName
```
Comment (epiqueras) [inline packages/edit-post/src/components/sidebar/settings-sidebar/index.js:31]:
```suggestion
		// component, besides being used to render the sidebar, also renders the toggle button. In that case sidebarName
```

File: packages/edit-post/src/components/sidebar/settings-sidebar/index.js [hunk 15]
```diff
@@ -22,25 +23,40 @@ import { __ } from '@wordpress/i18n';
...
+		// contains the sidebar that will be active when the toggle button is pressed. If a block
```
Comment (epiqueras) [inline packages/edit-post/src/components/sidebar/settings-sidebar/index.js:32]:
```suggestion
		// should contain the sidebar that will be active when the toggle button is pressed. If a block
```

File: packages/edit-post/src/components/sidebar/settings-sidebar/index.js [hunk 16]
```diff
@@ -22,25 +23,40 @@ import { __ } from '@wordpress/i18n';
...
+		// is selected the sidebar that will become active is edit-post/block otherwise is edit-post/document.
```
Comment (epiqueras) [inline packages/edit-post/src/components/sidebar/settings-sidebar/index.js:33]:
```suggestion
		// is selected, that should be edit-post/block otherwise it's edit-post/document.
```

File: packages/edit-site/src/components/sidebar/index.js [hunk 17]
```diff
@@ -6,13 +6,13 @@ import { ComplementaryArea } from '@wordpress/interface';
 import { __ } from '@wordpress/i18n';
 import { cog, pencil } from '@wordpress/icons';
 
-const { Slot: InspectorSlot, Fill: InspectorFill } = createSlotFill(
+const { Slot: InspectorSlot, Fill: Fill } = createSlotFill(
```
Comment (epiqueras) [inline packages/edit-site/src/components/sidebar/index.js:9]:
```suggestion
const { Slot: InspectorSlot, Fill: InspectorFill } = createSlotFill(
```

**Review by epiqueras — APPROVED** [review]
